python_modules/dagster/dagster/_core/definitions/asset_data_sla.py

```diff
@@ -0,0 +1,113 @@
+from typing import Mapping, Optional
+
+import pendulum
+from croniter import croniter
+
+from .events import AssetKey
+
+
+class RootDataTimeConstraint:
```

Comment:
This is clean. I like it.

Reply:
Thoughts on calling it `UpstreamMaterializationTimeConstraint`?
- It's conceivable that someone would want to set a constraint on an asset that's not a root asset.  E.g. if that asset depends on some upstream data but has an implementation that goes out and fetches some external data.
- "Materialization" has a clear meaning in the system, while "data" does not.

Reply:
Two more thoughts:
* This should be an abstract base class, right?
* In many (all?) cases, it's possible to calculate how late you are, and I suspect users would be curious about that. It might be useful to return that as well.

Reply:
Re: the non-root data times, I think this becomes pretty difficult to calculate, as to track this in the way we're currently doing (which allows O(1) look ups after the fact), you'd need to tag each asset with the data timestamp of every single asset upstream of it (rather than just the roots). Maybe that's actually fine in practice, but it feels pretty sketch to be adding potentially thousands of tags to each asset event.

For that specific scenario, I'd probably model that external thing that's being fetched from as a source asset (but maybe there are other scenarios in which you'd want to depend on a non-root asset's data where that wouldn't work?)

Agree on all other points

Reply:
> , as to track this in the way we're currently doing (which allows O(1) look ups after the fact), you'd need to tag each asset with the data timestamp of every single asset upstream of it (rather than just the roots)

I was imagining that the user would specify an SLA like "I don't want asset Y to be materialized later than 20 minutes after asset X", and then we would tag asset Y with the the timestamp of asset X, even if asset X was not a "root" asset.  Would that not work?